### DISCUSSION TITLE
Fetching RealT API

### DISCUSSION THREAD

**Discussion by nmathey** [MadeInPierre/finalynx#87] [Ideas]
Hello,

Comment pourrais-je faire pour récupérer mes investissements RealT via l['API actuelle](https://api.realt.community), le temps que l'équipe Finary fasse l'intégration en interne?

Tu parlais d'extension/plugin ? Tu as des exemples sur lesquels je pourrais m'appuyer?

J'ai quelque [bouts de codes](https://github.com/nmathey/RealT_OfflinePortfolio) que je pourrais réutiliser et même sans API Key on peut récupérer déjà une partie suffisante à mon sens pour la plupart des besoins.

**Comment by MadeInPierre** [comment]
Hello ! 

Idée très intéressante, pour l'instant Finalynx n'a rien de prévu pour intégrer d'autres sources que Finary mais ça pourrait être très puissant de le supporter :wink: 

Je pourrais adapter la partie `finalynx/fetch/` pour permettre des extensions d'API pour que n'importe qui puisse ajouter de nouvelles sources de deux manières : 
- Soit en ajoutant une source officielle au projet Finalynx pour que tout le monde puisse l'utiliser aussi dans le dossier `finalynx/fetch/fetch_mysource.py`,
- Soit en déclarant ta source ailleurs en local puis en la connectant à Finalynx au démarrage.

Pour ajouter RealT avec la 1ère option, tu pourrais créer un fichier `finalynx/fetch/fetch_realt.py` et je te laisse carte blanche pour te connecter à RealT comme tu préfères :smile: 

Ca te semble bien ? Si oui je m'y mets dans le prochains jours et tu pourras ajouter ta nouvelle source :slightly_smiling_face:

**Reply by sebfar9172** [reply]
Hello
En tout cas je suis intéressé par récupérer mes infos RealT (propriété et prix)

**Reply by nmathey** [reply]
> * Soit en ajoutant une source officielle au projet Finalynx pour que tout le monde puisse l'utiliser aussi dans le dossier `finalynx/fetch/fetch_mysource.py`,

Yes, je pense que cette option serait plus propre, il faudra juste bien me dire ce que doit retourner le tout et sous quel format :)

**Reply by MadeInPierre** [reply]
Issue créée pour suivre le développement (plus facile pour moi d'avoir des issues pour le dev plutôt qu'un mix issue/discussions) :
- #96

**Reply by MadeInPierre** [reply]
C'est implémenté dans #97, tu es libre de créer un nouveau fichier `finalynx/fetch/source_realt.py` en t'inspirant du [template exemple](https://github.com/MadeInPierre/finalynx/blob/main/finalynx/fetch/_source_template.py) et de la [source finary](https://github.com/MadeInPierre/finalynx/blob/main/finalynx/fetch/source_finary.py) ^^

N'hésite pas si tu as des questions ou idées d'amélioration !

**Reply by nmathey** [reply]
Yes j'ai commencé à travailler dessus mais dès mon premier test j'ai 
`ModuleNotFoundError: No module named 'finary_uapi'`

Je souhaite en plus juste lancé la source RealT que je viens de créer donc je suppose:
`python3 assistant_config.py --fetch-sources="realt"`

Je ne sais pas si c'est de ma faute en manipulant git de manière incorrect mais pourtant je suis bien parti de ta dernière branche main à jour...

**Reply by MadeInPierre** [reply]
Ah c'est peut-être à cause de #92, j'ai espoir que juste lancer `pip install finary_uapi` devrait résoudre le problème :crossed_fingers: 
S'il y a encore un problème c'est probablement que tu as une version python inférieure à 3.10, il faut mettre à jour python à cause de ce nouveau paquet...

**Comment by nmathey** [comment]
yep maintenant j'ai ca ^^

```
Usage:
  finalynx [ideal | targets | deltas | perf | text | --format=string] [--no-export | --export-dir=path] [dashboard] [options]
  finalynx (-h | --help)
  finalynx (-v | --version)
```

**Reply by MadeInPierre** [reply]
Désolé j'ai pas compris, ça ne marche pas ? S'il t'affiche le usage c'est qu'il y a un souci dans tes arguments d'entrée, t'as appelé finalynx comment ? Ceci devrait passer : 

```bash
python assistant_config.py --sources="finary,realt"
```

Normalement les options s'affichent avec `python assistant_config.py --help` ^^

**Reply by MadeInPierre** [reply]
Ahh j'ai changé de `--fetch-sources` à `--sources` sorry :sweat_smile: normalement ça passe avec `--sources`, voir avec `--help`

**Reply by nmathey** [reply]
Yes c'est mieux :) 
Maintenant il me dit qu'il ne trouve pas source: j'ai essayé 'realt' et 'RealT' ('SOURCE_NAME' dans mon fichier) mais 
```
Error: Source 'RealT' not recognized, have you added it   fetch.py:36
           first? Skipping.  
```

Je n'ai pas encore trouvé comment enregistrer ma source en lisant le code.

**Reply by MadeInPierre** [reply]
Ah, il manque ce morceau aussi dans mes explications >< 

Finalynx ne peut pas trouver tout seul qu'un nouveau fichier a été créé, il faut lui dire quelque part qu'il y a une nouvelle source.

Pour l'instant j'ai fait un truc simple (je préfère faire simple/moche mais facilement modifiable au début plutôt que "propre" mais en fait c'était pas ce qu'il fallait), tu peux ajouter ta source dans la [classe Assistant](https://github.com/MadeInPierre/finalynx/blob/6d39edcdb53d05711cf9c8d79ebac7f704071385/finalynx/assistant.py#L156-L157) avec un truc du genre :

```python
from finalynx.fetch.source_finary import SourceFinary
from finalynx.fetch.source_realt import SourceRealT

...

if "finary" in self.active_sources:
    self._fetch.add_source(SourceFinary(self.clear_cache, self.force_signin, self.ignore_orphans))
if "realt" in self.active_sources:
    self._fetch.add_source(SourceRealT(self.clear_cache, self.force_signin, self.ignore_orphans)) # et d'autres paramètres si besoin

...
```

S'il commence à y avoir pas mal de sources je ferai un système plus propre, voire détacherai tout ça dans un projet séparé (`finary_sync` ?) pour faire un "hub d'échange de données" que Finalynx consomme pour afficher/exploiter.

Normalement tu pourras ensuite appeler finalynx avec : 

```python
python assistant_config.py --sources="finary,realt"
```

avec `SOURCE_NAME = "RealT"` qui devrait fonctionner

**Reply by MadeInPierre** [reply]
D'ailleurs n'hésite pas à ouvrir une Pull Request avec des commits même si c'est pas fini, pas de soucis, ça me permettrait de voir l'état actuel et aider si besoin :slightly_smiling_face: on pourra merge quand ce sera fonctionnel

**Reply by nmathey** [reply]
J'arrivais justement au niveau de la classe Assistant dans ma lecture du code^^ Top j'avance la dessus et oui dès ca démarre à peu près je fais le PR même brouillon

**Reply by nmathey** [reply]
PR #106 créée.
J'ai laissé mon adresse publique Gnosis comportant mes RealToken si tu veux tester/améliorer le code (qui je pense peut être beaucoup mieux^^).

A ce stade pas besoin d'API Key pour récupérer plus de détails des bien, l'accès public à mon sens suffit.
En gros je récupère les infos de toutes la bases de biens RealT via l'API puis la liste de mes token sur mon wallet Gnosis pour ensuite garder que les Tokens REALTOKEN et armmREALTOKEN (tokens déposés en collateral sur RMM)

Y a peut-être un truc à faire aussi pour que le rendu soit plus propre: le numéro des addresses des biens ressortent en couleur comme la valeur détenue
